[Python][Flight] Show how to propagate opentelemetry spans to Flight (#299)
apache/arrow-cookbook: python/source/flight.rst

.. _IBM article: https://www.ibm.com/docs/en/arl/9.7?topic=certification-extracting-certificate-keys-from-pfx-file
.. _Windows: https://docs.microsoft.com/en-us/dotnet/core/additional-tools/self-signed-certificates-guide
.. _Arrow testing data repository: https://github.com/apache/arrow-testing/tree/master/data/flight
.. _openssl: https://www.ibm.com/docs/en/api-connect/2018.x?topic=overview-generating-self-signed-certificate-using-openssl

Propagating OpenTelemetry Traces
================================

> OpenTelemetry_ traces can be propagated between Flight clients and servers using

Reviewer: Minor thing but I might start this paragraph with the problem being solved with the example. Something like:

> In order to correlate spans across a Flight client server, trace context must be passed between the two. There are two methods to do this: (1) middleware (2) manually via .... The example shown below uses a Flight middleware (hyperlink) which has the advantage of separating the implementation of the client or server from injecting and extracting trace context.

> a middleware. The client middleware needs to inject the trace information into

Reviewer: ```suggestion
a middleware. The client middleware needs to inject the trace context into
```

I think we inject and extract "trace context" or "context" not spans. Do I have that right?

> the call headers. The server middleware needs to extract the trace information

Reviewer: ```suggestion
the call headers. The server middleware needs to extract the trace context
```

    from opentelemetry.trace.status import StatusCode

    class ServerTracingMiddlewareFactory(flight.ServerMiddlewareFactory):
        def __init__(self):
            self._tracer = trace.get_tracer(__name__)
        
        def start_call(self, info, headers):
            context = extract(headers)
            span = self._tracer.start_span(f"server.{info.method}", context=context)
            return ServerTracingMiddleware(span)
    
    class ServerTracingMiddleware(flight.ServerMiddleware):
        def __init__(self, span):
            self._span = span
        
        def call_completed(self, exception):
            if exception:
                self._span.record_exception(exception)
                self._span.set_status(StatusCode.ERROR)
                print(exception)
            else:
                self._span.set_status(StatusCode.OK)
            self._span.end()

Both the server and client will need to be configured with the OpenTelemetry SDK
to record spans and export them somewhere. For the sake of the example, we'll 
collect the spans into a Python list, but this is normally where you would set
them up to be exported to some service like `Jaeger`_.

> As part of this, you also define the resource: where the spans are running.

Reviewer: ```suggestion
As part of this, you will need to define the resource where spans are running.
```

        resource = Resource(attributes={
            SERVICE_NAME: "my-service"
        })
        exporter = TestSpanExporter()
        provider = TracerProvider(resource=resource)
        processor = SimpleSpanProcessor(exporter)
        provider.add_span_processor(processor)
        trace.set_tracer_provider(provider)
        return exporter

We can use the middleware now in our EchoServer from earlier. 

.. code-block::

    if __name__ == '__main__':
        exporter = configure_tracing()
        server = EchoServer(
            location="grpc://0.0.0.0:8816",
            middleware={
                "tracing": ServerTracingMiddlewareFactory()
            },
        )
        server.serve()

.. testcode::
    :hide:

    # Code block to start for real a server in background
    # and wait for it to be available.
    # Previous code block is just to show to user how to start it.
    import threading
    exporter = configure_tracing()
    server = EchoServer(
        location="grpc://0.0.0.0:8816",
        middleware={
            "tracing": ServerTracingMiddlewareFactory()
        },
    )
    t = threading.Thread(target=server.serve)
    t.start()

> Finally, we can use the middle on the client. When we make a call with our 

Reviewer: ```suggestion
Finally, we can use the middleware on the client. When we make a call with our 
```

> Flight client within an OpenTelemetry span, our client middleware will create

Reviewer: ```suggestion
client within an OpenTelemetry span, our client middleware will create
```

Keeping terminology consistent here ("client" and "server" have been used without "Flight" prepended)

> a child span for the client-side flight call and then propagate the span

Reviewer: ```suggestion
a child span for the client-side Flight call and then propagate the span
```

> context to the server. Our server middleware will pick up that context and create

Reviewer: "span context" or "trace context"? I want to say the latter.

Author: ```suggestion
context to the server. Our server middleware will pick up that trace context and create
```

.. _IBM article: https://www.ibm.com/docs/en/arl/9.7?topic=certification-extracting-certificate-keys-from-pfx-file
.. _Windows: https://docs.microsoft.com/en-us/dotnet/core/additional-tools/self-signed-certificates-guide
.. _Arrow testing data repository: https://github.com/apache/arrow-testing/tree/master/data/flight
.. _openssl: https://www.ibm.com/docs/en/api-connect/2018.x?topic=overview-generating-self-signed-certificate-using-openssl

Propagating OpenTelemetry Traces
================================

Distributed tracing with OpenTelemetry_ allows collecting call-level performance
measurements across a Flight service. In order to correlate spans across a Flight

> client and server, trace context must be passed between the two. This can be passed

Reviewer 2: Are we OK with using jargon without (briefly) defining it?

Author: I think so. My interpretation of the Cookbook is that it's mainly for sharing ready to use examples, not walking the reader through all the concepts. I think by the time the user is googling "arrow flight opentelemetry python" they know what these terms mean.

Reviewer 2: OK, sounds good to me.

    t.start()

Finally, we can use the middleware on the client. When we make a call with our 
client within an OpenTelemetry span, our client middleware will create
a child span for the client-side Flight call and then propagate the span
context to the server. Our server middleware will pick up that trace context and create
another child span.

.. testcode::

   import pyarrow as pa
   import pyarrow.flight
   from opentelemetry import trace

   # Client would normally also need to configure tracing, but for this example
   # the client and server are running in the same Python process.
   # exporter = configure_tracing()

   tracer = trace.get_tracer(__name__)

   client = pa.flight.connect(
       "grpc://0.0.0.0:8816",
       middleware=[ClientTracingMiddlewareFactory()],
   )

   with tracer.start_as_current_span("hello_world") as span:
       action = pa.flight.Action("echo", b"Hello, world!")
       # Call list() on do_action to drain all results.
       list(client.do_action(action=action))
    
   print(f"There are {len(exporter.spans)} spans.")
   print(f"The span names are:\n  {list(span.name for span in exporter.spans)}.")
   print(f"The span status codes are:\n  "

>          f"{list(span.status.status_code for span in exporter.spans)}.")

Reviewer 2: It may not be clear here, but the status codes here are OpenTelemetry statuses, and not gRPC or Flight statuses

Author: We set them in the middleware above, so I think that's self explanatory enough for now. I suppose a more full-featured middleware would also set the gRPC status in the middleware with semantic conventions, but that seemed like it would overcomplicate this.
